srcts/src/shiny/bind.ts

```diff
@@ -61,31 +61,32 @@ const bindingsRegistery = (() => {
   const bindings: IdToBindingTypes = new Map();
 
   /**
-   * Checks if the bindings registery is valid. Currently this just checks for
+   * Checks if the bindings registry is valid. Currently this just checks for
    * duplicate IDs but in the future could be expanded to check more conditions
-   * @returns ShinyClientError if current ID bindings are invalid, otherwise null
+   * @returns ShinyClientError if current ID bindings are invalid, otherwise
+   * returns an ok status.
    */
   function checkValidity():
     | { status: "error"; error: ShinyClientError }
     | { status: "ok" } {
-    const duplicateIds: IdToBindingTypes = new Map();
+    const duplicateIds: { [id: string]: { input: number; output: number } } =
+      {};
 
-    bindings.forEach((inputOrOutput, id) => {
-      if (inputOrOutput.length > 1) {
-        duplicateIds.set(id, inputOrOutput);
+    // count duplicate IDs of each binding type
+    bindings.forEach((idTypes, id) => {
+      const nInputs = idTypes.filter((s) => s === "input").length;
+      const nOutputs = idTypes.filter((s) => s === "output").length;
+
+      if (nInputs > 1 || nOutputs > 1) {
+        duplicateIds[id] = { input: nInputs, output: nOutputs };
       }
```

Comment:
Is this the only bit of logic that's different here? If it is I may advocate for reducing changes to just this to keep changes focused. The new code feels a bit harder to follow (although that's probably because I wrote the original.)

Reply:
Sort of, but we also have to count the number of each type earlier. Before you would add an `id` to `duplicateIds` if `bindings.get(id).length > 1` but we need this logic to happen earlier so we can take into account whether there are more than one of the same type as the inclusion criteria.